pkg/controller/sync/placement.go

```diff
@@ -0,0 +1,115 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+)
+
+// computeNamespacedPlacement determines placement for namespaced
+// federated resources (e.g. FederatedConfigMap).
+//
+// If federation is deployed cluster-wide, placement is the
+// intersection of the placement for the federated resource and the
+// placement of the federated namespace containing the resource.
+//
+// If federation is limited to a single namespace, placement is
+// determined as the intersection of resource and namespace placement
+// if namespace placement exists.  If namespace placement does not
+// exist, resource placement will be used verbatim.  This is possible
+// because the single namespace by definition must exist on member
+// clusters, so namespace placement becomes a mechanism for limiting
+// rather than allowing propagation.
+func computeNamespacedPlacement(resource, namespace *unstructured.Unstructured, clusters []*fedv1a1.FederatedCluster, limitedScope bool) (selectedClusters, unselectedClusters []string, err error) {
+	selectedClusters, unselectedClusters, err = computePlacement(resource, clusters)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	clusterNames := getClusterNames(clusters)
+
+	if namespace == nil {
+		if limitedScope {
+			// Use the resource placement verbatim if no federated
+			// namespace is present and federation is targeting a
+			// single namespace.
+			return selectedClusters, unselectedClusters, nil
+		}
+		// Resource should not exist in any member clusters.
+		return []string{}, clusterNames, nil
+	}
+
+	namespaceSelectedClusters, _, err := computePlacement(namespace, clusters)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resourceClusterSet := sets.NewString(selectedClusters...)
+	namespaceClusterSet := sets.NewString(namespaceSelectedClusters...)
+	clusterSet := sets.NewString(clusterNames...)
+
+	// If both namespace and resource placement exist, the desired
+	// list of clusters is their intersection.
+	selectedSet := resourceClusterSet.Intersection(namespaceClusterSet)
+
+	return selectedSet.List(), clusterSet.Difference(selectedSet).List(), nil
+}
+
+// computePlacement determines the selected and unselected clusters
+// for a federated resource.
+func computePlacement(resource *unstructured.Unstructured, clusters []*fedv1a1.FederatedCluster) (selectedClusters, unselectedClusters []string, err error) {
+	selectedNames, err := selectedClusterNames(resource, clusters)
+	if err != nil {
+		return nil, nil, err
+	}
+	clusterNames := getClusterNames(clusters)
+	clusterSet := sets.NewString(clusterNames...)
+	selectedSet := sets.NewString(selectedNames...)
+	return clusterSet.Intersection(selectedSet).List(), clusterSet.Difference(selectedSet).List(), nil
+}
+
+func selectedClusterNames(rawPlacement *unstructured.Unstructured, clusters []*fedv1a1.FederatedCluster) ([]string, error) {
```

Comment:
s/rawPlacement/rawResource/

Reply:
Done

Reply:
I've changed `rawPlacement` to `resource` for consistency with the other functions in this file.

Reply:
> @shashidharatd @irfanurrehman Thoughts on a new group name to replace `primitives.federation.k8s.io`?

Reasons for thinking on this change? `primitives` still sounds quite ok to me. `core`, `base`, `primary` are the other options that come to my mind.